File: clients/typescript/test/live-teardown.test.ts
import { describe, it, expect, vi } from "vitest";
import { LiveCollection } from "../src/live/live-collection.js";
import type { RealtimeCallback, RealtimeEvent } from "../src/realtime.js";

/** Realtime stub that records unsubscribe invocations and the returned unsub fns. */
function fakeRealtime() {
  const subs = new Map<string, Set<RealtimeCallback>>();
  const unsubscribeCalls: Array<{ topic: string; cb?: RealtimeCallback }> = [];
  let unsubFnCalls = 0;
  return {
    service: {
      async subscribe(topic: string, cb: RealtimeCallback) {
        let set = subs.get(topic);
        if (!set) subs.set(topic, (set = new Set()));
        set.add(cb);
        return () => {
          unsubFnCalls += 1;
          set!.delete(cb);
        };
      },
      unsubscribe(topic: string, cb?: RealtimeCallback) {
        unsubscribeCalls.push({ topic, cb });
        if (cb) subs.get(topic)?.delete(cb);
        else subs.delete(topic);
      },
    },
    emit(topic: string, event: RealtimeEvent) {
      for (const cb of subs.get(topic) ?? []) cb(event);
    },
    subscriberCount(topic: string) {
      return subs.get(topic)?.size ?? 0;
    },
    get unsubFnCalls() {
      return unsubFnCalls;
    },
  };
}

describe("LiveList.close teardown", () => {
  it("unsubscribes from realtime and releases every retained cache ref", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({
        items: [
          { id: "a", rank: 1 },
          { id: "b", rank: 2 },
        ],
        page: 1,
        perPage: 30,
        totalItems: 2,
      })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);
    const list = await lc.getList(1, 30, { sort: "rank" });
    expect(rt.subscriberCount("posts")).toBe(1);

    list.close();

    // Realtime unsubscribe was invoked and the topic callback removed.
    expect(rt.unsubFnCalls).toBeGreaterThanOrEqual(1);
    expect(rt.subscriberCount("posts")).toBe(0);

    // After teardown the list is emptied and events no longer mutate it.
    expect(list.items).toEqual([]);
    rt.emit("posts", { topic: "posts", action: "create", record: { id: "c", rank: 0 } });
    expect(list.items).toEqual([]);

    // The underlying cache entries were released (refcount -> 0, evicted).
    expect((lc as unknown as { cache: { has(id: string): boolean } }).cache.has("a")).toBe(false);
    expect((lc as unknown as { cache: { has(id: string): boolean } }).cache.has("b")).toBe(false);
  });

  it("is idempotent: a second close() does not double-free", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({
        items: [{ id: "a", rank: 1 }],
        page: 1,
        perPage: 30,
        totalItems: 1,
      })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);
    const list = await lc.getList(1, 30, { sort: "rank" });

    list.close();
    const before = rt.unsubFnCalls;
    list.close(); // no-op
    expect(rt.unsubFnCalls).toBe(before);
    // Cache ref for "a" stays at 0 (not driven negative).
    expect((lc as unknown as { cache: { has(id: string): boolean } }).cache.has("a")).toBe(false);
  });

  it("stops a pending refetch timer on close", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({
        items: [{ id: "a", rank: 1 }],
        page: 1,
        perPage: 30,
        totalItems: 1,
      })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);
    let scheduled: (() => void) | null = null;
    const cleared: unknown[] = [];
    const list = await lc.getList(1, 30, {
      filter: "author.name = 'Ada'", // relation -> refetch tier
      sort: "rank",
      schedule: (fn) => {
        scheduled = fn;
        return 123 as never;
      },
    });
    rt.emit("posts", { topic: "posts", action: "create", record: { id: "b", rank: 2 } });
    expect(typeof scheduled).toBe("function");
    // close should clear the timer; we just assert no throw and no refetch after.
    const calls = reader.getList.mock.calls.length;
    list.close();
    await new Promise((r) => setTimeout(r, 0));
    expect(reader.getList.mock.calls.length).toBe(calls);
    void cleared;
  });
});

describe("LiveRecord.close teardown (getOne)", () => {
  it("unsubscribes the name/id topic and releases its cache ref", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(async (id: string) => ({ id, title: "Seed" })),
      getList: vi.fn(),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);
    const live = await lc.getOne("p1");
    expect(rt.subscriberCount("posts/p1")).toBe(1);

    live.close();
    expect(rt.unsubFnCalls).toBeGreaterThanOrEqual(1);
    expect(rt.subscriberCount("posts/p1")).toBe(0);
    expect((lc as unknown as { cache: { has(id: string): boolean } }).cache.has("p1")).toBe(false);

    // Idempotent.
    const before = rt.unsubFnCalls;
    live.close();
    expect(rt.unsubFnCalls).toBe(before);
  });
});

describe("LiveList.mode DX surface", () => {
  it("reports 'precise' for an own-field filter and 'refetch' for a relation filter", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({ items: [], page: 1, perPage: 30, totalItems: 0 })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);

    const precise = await lc.getList(1, 30, { filter: "status = 'published'", sort: "rank" });
    expect(precise.mode).toBe("precise");

    const refetch = await lc.getList(1, 30, { filter: "author.name = 'Ada'", sort: "rank" });
    expect(refetch.mode).toBe("refetch");

    const noFilter = await lc.getList(1, 30, { sort: "rank" });
    expect(noFilter.mode).toBe("precise");
  });
});

File: clients/dart/test/live/live_teardown_test.dart
import 'package:test/test.dart';
import 'package:zigbase_client/zigbase_client.dart';

import 'fakes.dart';

void main() {
  group('LiveList.close teardown', () {
    test('unsubscribes from realtime and releases every retained cache ref',
        () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'rank': 1}),
              rec({'id': 'b', 'rank': 2}),
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');
      expect(rt.subscriberCount('posts'), 1);

      list.close();

      // Realtime unsubscribe was invoked and the topic callback removed.
      expect(rt.unsubFnCalls, greaterThanOrEqualTo(1));
      expect(rt.subscriberCount('posts'), 0);

      // The underlying cache entries were released (refcount -> 0, evicted).
      expect(lc.cache.has('a'), isFalse);
      expect(lc.cache.has('b'), isFalse);

      // After teardown the list is terminal: reads throw, events are ignored.
      expect(() => list.items, throwsStateError);
      rt.emit(
          'posts', RecordEvent('posts', 'create', rec({'id': 'c', 'rank': 0})));
      expect(lc.cache.has('c'), isFalse);
    });

    test('is idempotent: a second close() does not double-free', () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'rank': 1})
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');

      list.close();
      final before = rt.unsubFnCalls;
      list.close(); // no-op
      expect(rt.unsubFnCalls, before);
      expect(lc.cache.has('a'), isFalse);
    });

    test('stops a pending refetch timer on close', () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'rank': 1})
            ];
      final lc = LiveCollection('posts', reader, rt);
      void Function()? scheduled;
      var cleared = 0;
      final list = await lc.getList(
        filter: "author.name = 'Ada'", // relation -> refetch tier
        sort: 'rank',
        schedule: (run, delay) {
          scheduled = run;
          return Object();
        },
        clearSchedule: (handle) => cleared += 1,
      );
      rt.emit(
          'posts', RecordEvent('posts', 'create', rec({'id': 'b', 'rank': 2})));
      expect(scheduled, isNotNull);

      final calls = reader.listCalls;
      list.close();
      await pumpEventQueue();
      // The pending timer was cleared and no refetch fired.
      expect(cleared, 1);
      expect(reader.listCalls, calls);
    });
  });

  group('LiveList.mode DX surface', () {
    test('reports precise for an own-field filter and refetch for a relation',
        () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()..onGetListItems = () async => [];
      final lc = LiveCollection('posts', reader, rt);

      final precise =
          await lc.getList(filter: "status = 'published'", sort: 'rank');
      expect(precise.mode, LiveListMode.precise);

      final refetch =
          await lc.getList(filter: "author.name = 'Ada'", sort: 'rank');
      expect(refetch.mode, LiveListMode.refetch);

      final noFilter = await lc.getList(sort: 'rank');
      expect(noFilter.mode, LiveListMode.precise);
    });
  });
}
